builtins.go

```diff
@@ -17,200 +17,290 @@
 		p, err := strconv.ParseInt(param, 0, 64)
 		if err != nil {
-			return nil, ErrBadParameter
+			return nil, errBadParameter
 		}
-		return func(v reflect.Value, state *validateState) {
+		return func(v reflect.Value, r *ErrorReporter) {
 			if int64(v.Len()) > p {
-				state.error(ErrMax)
+				r.Errorf(errMax)
 			}
 		}, nil
 	case reflect.Float32, reflect.Float64:
 		p, err := strconv.ParseFloat(param, 64)
 		if err != nil {
-			return nil, ErrBadParameter
+			return nil, errBadParameter
 		}
-		return func(v reflect.Value, state *validateState) {
+		return func(v reflect.Value, r *ErrorReporter) {
 			if v.Float() > p {
-				state.error(ErrMax)
+				r.Errorf(errMax)
 			}
 		}, nil
 	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
 		p, err := strconv.ParseInt(param, 0, 64)
 		if err != nil {
-			return nil, ErrBadParameter
+			return nil, errBadParameter
 		}
-		return func(v reflect.Value, state *validateState) {
+		return func(v reflect.Value, r *ErrorReporter) {
 			if v.Int() > p {
-				state.error(ErrMax)
+				r.Errorf(errMax)
+			}
+		}, nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		p, err := strconv.ParseUint(param, 0, 64)
+		if err != nil {
+			return nil, errBadParameter
+		}
+		return func(v reflect.Value, r *ErrorReporter) {
+			if v.Uint() > p {
+				r.Errorf(errMax)
 			}
 		}, nil
 	default:
-		return nil, ErrUnsupported
+		return nil, errUnsupportedType
 	}
 }
 
 // regex is the builtin validation function that checks
 // whether the string variable matches a regular expression
-func regex(t reflect.Type, param string) (validationFunc, error) {
+func regex(t reflect.Type, param string) (ValidationFunc, error) {
 	re, err := regexp.Compile(param)
 	if err != nil {
-		return nil, ErrBadParameter
+		return nil, errBadParameter
 	}
 	if t != reflect.TypeOf("") {
-		return nil, ErrUnsupported
+		return nil, errUnsupportedType
 	}
-	return func(v reflect.Value, state *validateState) {
+	return func(v reflect.Value, r *ErrorReporter) {
 		if !re.MatchString(v.String()) {
-			state.error(ErrRegexp)
+			r.Errorf(errRegexp)
 		}
 	}, nil
 }
+
+// required validates the value is not nil for a field, that is, a
+// pointer or an interface, any other case is a valid one as zero
+// value from Go spec
+func required(t reflect.Type, param string) (ValidationFunc, error) {
+	switch t.Kind() {
+	case reflect.Ptr, reflect.Interface:
+		return func(v reflect.Value, r *ErrorReporter) {
+			if v.IsNil() {
+				r.Errorf(errRequired)
+			}
+		}, nil
+	case reflect.String, reflect.Slice, reflect.Map, reflect.Array, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr, reflect.Float32, reflect.Float64, reflect.Bool, reflect.Struct:
+		// TODO always validating OK for slice and map types seems like
```

Comment:
I don't want to risk breaking existing code in universe that's reliant on these semantics. I don't believe there's anything that's using the `len=` validation for numeric types, but there are plenty of places that are using `required` on strings or ints, though I suspect they're almost all done by mistake. An arbitrary example:

```
 	TotalPrice         int              `json:"total_price" validate:"required"`
```

The `validate` tag isn't doing anything in this case. If we change the semantics now and we're actually using zero `TotalPrice` values, then this will break production.

This seems like too big of a risk to me right now - I want to make something that is essentially risk-free for people to upgrade to. We can make a v5 later if we want to change this.

Reply:
Yeah, they are all wrong, I am fixing those in some endpoints in universe. I was talking about Maps and Slices that they seem right to me.